buildRustCrate + strip produces invalid rlib on Darwin

# tl;dr

Darwin's `strip` v973 appears to break rust `.rlib` files. The previous shipped version (v949) didn't break anything, but that may have been accidental.

# Reproduction:

```nix
# default.nix
with import (builtins.fetchGit {
        url = "https://github.com/NixOS/nixpkgs.git";
        rev = "e3945057be467f32028ff6b67403be08285ad8c8";
}) {};

let
cfg_if = buildRustCrate {
        crateName = "cfg_if";
        version = "1.0.0";
        dontStrip = false;
        src = fetchurl {
                hash = "sha256-uvHeQzl2FYi8Bhnjy8ASDuWC67dLU7Tvv3kRe9LaQP0=";
                name = "crate.tar.gz";
                url = "https://crates.io/api/v1/crates/cfg-if/1.0.0/download";
        };
        edition = "2018";
};

encoding = buildRustCrate {
        crateName = "encoding_rs";
        version = "0.8.32";
        dependencies = [ cfg_if ];
        src = fetchurl {
                hash = "sha256-Bxox9O6FQDNwtYrKdG8BBB7ebw2icwlgrQAe3Ctxs5Q=";
                name = "crate.tar.gz";
                url = "https://crates.io/api/v1/crates/encoding_rs/0.8.32/download";
        };
        edition = "2018";
        features = [ "alloc" "default" ];
};

in encoding
```

```
$ nix-build
# ...
error[E0786]: found invalid metadata files for crate `cfg_if`
   --> src/lib.rs:700:1
    |
700 | extern crate cfg_if;
    | ^^^^^^^^^^^^^^^^^^^^
    |
    = note: no `.rmeta` section in '/nix/store/z2ann7q20qfzipfr87cqm3qwmli7i3zd-rust_cfg_if-1.0.0-lib/lib/libcfg_if-7ec8491762.rlib'

# ... followed by lots of unresolved symbol errors
```

If you set `dontStrip = true;` in the cfg-if derivation, the build succeeds. The resulting `encoding` rlib file is likely invalid too, this is just a minimal reproduction.

### On linux:

The same derivation builds fine:

/nix/store/x5lz2zrq2xkj02xn7mkxm19nknik202k-rust_encoding_rs-0.8.32

So presumably there's some difference in how `strip` works.

### Narrowing down `nixpkgs` versions

`git bisect` of `nixpkgs` got me close, but many commits needed skipping due to unrelated issues. The best I could get is:

 - Oldest working commit: 6d1f3e441e27483beca8e504d471fb7865e5fb52 (oct 1 2022)
 - Earliest failing commit: ad9975e5de6b396744ffb897ee7eb88c8800b10e (oct 20 2022)

The most relevant looking commit in that range is:

```
71b05eccaaf4dad1f01b13832cb0d692b8e1e761 darwin.cctools: 949.0.1 -> 973.0.1
```

The `darwin.cctools` derivation from the above nixpkgs commits are:
 - /nix/store/a5ac77k6289sgbq5izv66djgsvabzhc4-cctools-port-973.0.1
 - /nix/store/mz472jcxzy73rw8297zabcczns3gmjf6-cctools-port-949.0.1

# The cause

Assuming that's the culprit, I copied the non-stripped `rlib` and manually ran `strip -S` using both these implementations. This confirms that v973 appears to be the culprit - the v949 version results in a `.rmeta` section while v973 removes it.

I note that the strip command prints `warning: input object file already stripped`. Based on [this change in cctools-port](https://github.com/tpoechtrager/cctools-port/commit/86cf74fd50016550703435302c33c17474759666#diff-8d3878ab2b0e84f0b73e9eae9f8688b492758b6bb12632972799fbac670895beR1114-R1118), it looks like the earlier version bails and essentially does nothing, while the later version proceeds to do more stripping on an already-stripped file, resulting in the .rmeta section being removed.

# Fix options:

 - should we never strip `rlib` files? It says they're already stripped, maybe stripping isn't relevant to this kind of file?
 - is it a bug that darwin's `strip` removes the `.rmeta` section, and this should be perfectly fine?
 - ... other options?

<details><summary>
<h2>Inspection details</h2>
</sumary>

<pre>
# extract meta file
$ ar -p "/nix/store/z2ann7q20qfzipfr87cqm3qwmli7i3zd-rust_cfg_if-1.0.0-lib/lib/libcfg_if-7ec8491762.rlib" lib.rmeta > /tmp/lib.rmeta

# dump rmeta sections
$ otool -l /tmp/lib.rmeta

/tmp/lib.rmeta:
Load command 0
      cmd LC_SEGMENT_64
  cmdsize 72
  segname 
   vmaddr 0x0000000000000000
   vmsize 0x0000000000000000
  fileoff 0
 filesize 0
  maxprot 0x00000007
 initprot 0x00000007
   nsects 0
    flags 0x0
Load command 1
     cmd LC_SYMTAB
 cmdsize 24
  symoff 0
   nsyms 0
  stroff 176
 strsize 8
Load command 2
      cmd LC_DATA_IN_CODE
  cmdsize 16
  dataoff 176
 datasize 0
</pre>

Note there's no `.rmeta` section. With `dontStrip = true` to the cfg-if derivation, I get:

<pre>
/tmp/lib.rmeta:
Load command 0
      cmd LC_SEGMENT_64
  cmdsize 152
  segname 
   vmaddr 0x0000000000000000
   vmsize 0x0000000000001883
  fileoff 208
 filesize 6275
  maxprot 0x00000007
 initprot 0x00000007
   nsects 1
    flags 0x0
Section
  sectname .rmeta
   segname __DWARF
      addr 0x0000000000000000
      size 0x0000000000001883
    offset 208
     align 2^0 (1)
    reloff 0
    nreloc 0
     flags 0x02000000
 reserved1 0
 reserved2 0
Load command 1
     cmd LC_SYMTAB
 cmdsize 24
  symoff 6488
   nsyms 0
  stroff 6488
 strsize 1
 </pre>
</details>

--- a/doc/stdenv/stdenv.chapter.md
+++ b/doc/stdenv/stdenv.chapter.md
@@ -937,6 +937,28 @@ Like `stripDebugList`, but only applies to packages’ target platform. By defau
 
 Flags passed to the `strip` command applied to the files in the directories listed in `stripDebugList`. Defaults to `-S` (i.e. `--strip-debug`).
 
+##### `stripExclude` {#var-stdenv-stripExclude}
+
+A list of filenames or path patterns to avoid stripping. A file is excluded if its name _or_ path (from the derivation root) matches.
+
+This example prevents all `*.rlib` files from being stripped:
+
+```nix
+stdenv.mkDerivation {
+  # ...
+  stripExclude = [ "*.rlib" ]
+}
+```
+
+This example prevents files within certain paths from being stripped:
+
+```nix
+stdenv.mkDerivation {
+  # ...
+  stripExclude = [ "lib/modules/*/build/* ]
+}
+```
+
 ##### `dontPatchELF` {#var-stdenv-dontPatchELF}
 
 If set, the `patchelf` command is not used to remove unnecessary `RPATH` entries. Only applies to Linux.

--- a/pkgs/build-support/setup-hooks/strip.sh
+++ b/pkgs/build-support/setup-hooks/strip.sh
@@ -49,10 +49,18 @@ stripDirs() {
     local ranlibCmd="$2"
     local paths="$3"
     local stripFlags="$4"
+    local excludeFlags=()
     local pathsNew=
 
     [ -z "$cmd" ] && echo "stripDirs: Strip command is empty" 1>&2 && exit 1
     [ -z "$ranlibCmd" ] && echo "stripDirs: Ranlib command is empty" 1>&2 && exit 1
+
+    local pattern
+    if [ -n "${stripExclude:-}" ]; then
+        for pattern in "${stripExclude[@]}"; do
+            excludeFlags+=(-a '!' '(' -name "$pattern" -o -wholename "$prefix/$pattern" ')' )
+        done
+    fi
 
     local p
     for p in ${paths}; do
@@ -67,7 +75,7 @@ stripDirs() {
         local striperr
         striperr="$(mktemp --tmpdir="$TMPDIR" 'striperr.XXXXXX')"
         # Do not strip lib/debug. This is a directory used by setup-hooks/separate-debug-info.sh.
-        find $paths -type f -a '!' -path "$prefix/lib/debug/*" -print0 |
+        find $paths -type f "${excludeFlags[@]}" -a '!' -path "$prefix/lib/debug/*" -print0 |
             # Make sure we process files under symlinks only once. Otherwise
             # 'strip` can corrupt files when writes to them in parallel:
             #   https://github.com/NixOS/nixpkgs/issues/246147#issuecomment-1657072039

